From: Hongjin Li <vic.hong@rock-chips.com>
Date: Wed, 27 Nov 2024 17:25:03 +0800
Subject: fix[av1_vdpu383]: fix cdf usage issue

Platform: rk3576/vdpu383

Error case:
The idr frame (frm0) does not update the cdf, and then
frm1 uses the invalid cdf of frm0, causing core timeout

Solution:
Add a flag to determine whether the cdf is valid

From: BSP Platform Department yk

Source: av1.rk3576.failure.obu

Signed-off-by: Hongjin Li <vic.hong@rock-chips.com>
Change-Id: I72d08877d40368976794ba3bcabf4e939d4fa047
---
 mpp/hal/rkdec/av1d/hal_av1d_vdpu383.c | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

diff --git a/mpp/hal/rkdec/av1d/hal_av1d_vdpu383.c b/mpp/hal/rkdec/av1d/hal_av1d_vdpu383.c
index 4978cf2..ee7b171 100644
--- a/mpp/hal/rkdec/av1d/hal_av1d_vdpu383.c
+++ b/mpp/hal/rkdec/av1d/hal_av1d_vdpu383.c
@@ -136,6 +136,7 @@ typedef struct vcpu383_ref_info_t {
     RK_U32 dpb_idx;
     RK_U32 seg_idx;
     RK_U32 colmv_exist_flag;
+    RK_U32 cdf_valid;
     RK_U32 coeff_idx;
     RK_U32 mi_rows;
     RK_U32 mi_cols;
@@ -2065,16 +2066,13 @@ static void vdpu383_av1d_set_cdf(Av1dHalCtx *p_hal, DXVA_PicParams_AV1 *dxva)
         sprintf(dump_cur_fname_path, "%s/%s", dump_cur_dir, cur_fname);
     }
 #endif
-    if (dxva->format.frame_type == AV1_FRAME_KEY || dxva->primary_ref_frame == 7 /* AV1_PRIMARY_REF_NONE */) {
-        if (dxva->quantization.base_qindex <= 20)
-            coeff_cdf_idx = 0;
-        else if (dxva->quantization.base_qindex <= 60)
-            coeff_cdf_idx = 1;
-        else if (dxva->quantization.base_qindex <= 120)
-            coeff_cdf_idx = 2;
-        else
-            coeff_cdf_idx = 3;
+    /* def coeff cdf idx */
+    coeff_cdf_idx = dxva->quantization.base_qindex <= 20 ? 0 :
+                    dxva->quantization.base_qindex <= 60  ? 1 :
+                    dxva->quantization.base_qindex <= 120 ? 2 : 3;
 
+    if (dxva->format.frame_type == AV1_FRAME_KEY ||
+        dxva->primary_ref_frame == 7) { /* AV1_PRIMARY_REF_NONE */
         regs->av1d_addrs.reg184_av1_noncoef_rd_base = mpp_buffer_get_fd(reg_ctx->cdf_rd_def_base);
         regs->av1d_addrs.reg178_av1_coef_rd_base = mpp_buffer_get_fd(reg_ctx->cdf_rd_def_base);
 #ifdef DUMP_AV1D_VDPU383_DATAS
@@ -2090,7 +2088,8 @@ static void vdpu383_av1d_set_cdf(Av1dHalCtx *p_hal, DXVA_PicParams_AV1 *dxva)
         mapped_idx = dxva->ref_frame_idx[dxva->primary_ref_frame];
 
         coeff_cdf_idx = reg_ctx->ref_info_tbl[mapped_idx].coeff_idx;
-        if (!dxva->coding.disable_frame_end_update_cdf) {
+        if (!dxva->coding.disable_frame_end_update_cdf &&
+            reg_ctx->ref_info_tbl[mapped_idx].cdf_valid) {
             cdf_buf = hal_bufs_get_buf(reg_ctx->cdf_bufs, dxva->frame_refs[mapped_idx].Index);
             buf_tmp = cdf_buf->buf[0];
         } else {
@@ -2130,6 +2129,7 @@ static void vdpu383_av1d_set_cdf(Av1dHalCtx *p_hal, DXVA_PicParams_AV1 *dxva)
                 else
                     reg_ctx->ref_info_tbl[i].coeff_idx = coeff_cdf_idx;
             } else {
+                reg_ctx->ref_info_tbl[i].cdf_valid = 1;
                 reg_ctx->ref_info_tbl[i].coeff_idx = 0;
             }
         }
